File: Course_2-Data_Structures/Week-4/Excercise_Challenges/2_hash_chains/hash_chains.py
class HashChain:
    def __init__(self):
        self.m = 0
        self.multiplier = 263
        self.mod = 1000000007
        self.data = {}

        
    def hash_func(self, s):
        v = 0
        for i in range(len(s)):
            v += (ord(s[i]) * (self.multiplier ** i))
            v %= self.mod
        
        return v % self.m


    def process_query(self, s):
        if s[0] == 'add':
            temp = self.hash_func(s[1])
            
            if temp not in self.data:
                self.data[temp] = [s[1]]
            elif temp in self.data and s[1] not in self.data[temp]:
                self.data[temp].append(s[1])
        
        elif s[0] == 'check':
            temp = int(s[1])

            if temp in self.data:
                for v in self.data[temp][::-1]:
                    print(v, end=" ")
                print()
            else:
                print()
            
        elif s[0] == 'find':
            temp = self.hash_func(s[1])

            if temp not in self.data: 
                print("no")
            else:
                if s[1] not in self.data[temp]:
                    print("no")
                else:
                    print("yes")
                
        elif s[0] == 'del':
            temp = self.hash_func(s[1])

            if temp in self.data and s[1] in self.data[temp]:
                self.data[temp].remove(s[1])

    def read_query(self):
        self.m = int(input())
        q = int(input())

        for _ in range(q):
            s = input().split()
            self.process_query(s)

    def solve(self):
        self.read_query()

if __name__ == '__main__':
    obj = HashChain()
    obj.solve()

File: Course_2-Data_Structures/Week-4/Excercise_Challenges/2_hash_chains/hash_chains.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <cmath>
#include <map>
#define MOD 1000000007
typedef long long ll;
using namespace std;


int hash_func(string a, int m)
{
    int len = a.length();

    ll v = 0;
    for(int i = len - 1; i >= 0; --i){
        v = (v * 263 + (int)a[i]) % MOD;
    }

    return v % m; 
}

bool search(string s, vector<string> v)
{
    for(auto a : v){
        if(a == s)
            return true;
    }
    return false;
}

int main(void)
{
    std::ios_base::sync_with_stdio(false);
    int n;
    cin >> n;
    int q;
    cin >> q;

    string a, b;
    int temp;

    map<int, vector<string>> m;
    for(int i = 0; i < q; ++i){
        cin >> a;

        if(a == "add"){
            cin >> b;

            temp = hash_func(b, n);
            bool flag = search(b, m[temp]);
            if(!flag)
                m[temp].push_back(b);
        }
        else if(a == "check"){
            cin >> temp;

            if(m.count(temp)){
                for(int j = m[temp].size() - 1; j >= 0; --j)
                    cout << m[temp][j] << " ";
                cout << endl;
            }
            else
                cout << endl;
        }
        else if(a == "find"){
            cin >> b;

            temp = hash_func(b, n);
            if(m.count(temp)){
                bool flag = search(b, m[temp]);

                if(flag)
                    cout << "yes" << endl;
                else
                    cout << "no" << endl;
            }
            else{
                cout << "no" << endl;
            }
        }
        else if(a == "del"){
            cin >> b;

            temp = hash_func(b, n);
            if(m.count(temp)){
                int nlen = m[temp].size();
                for(int j = 0; j < nlen; ++j){
                    if(m[temp][j] == b){
                        m[temp].erase(m[temp].begin() + j);
                        break;
                    }
                }
            }
        }
    }
    

    return 0;
}
